src/viam/ur/module/ur_arm.cpp

```diff
@@ -172,128 +184,1284 @@ void write_waypoints_to_csv(const std::string& filepath, const std::list<Eigen::
+        for (size_t i = 0; i < num_samples; i++) {
+            if (!arm_conn_->driver->writeTrajectorySplinePoint(samples[i].p, samples[i].v, samples[i].timestep)) {
+                VIAM_SDK_LOG(error) << "send_trajectory cubic driver->writeTrajectorySplinePoint returned false";
+                std::exchange(state.move_request_, {})->complete_error("failed to send trajectory spline point to arm");
+                return event_local_mode_detected_{};
+            };
+        }
+
+        VIAM_SDK_LOG(info) << "URArm trajectory sent";
+
+    } else if (state.move_request_->samples.empty() && state.move_request_->cancellation_request &&
+               !state.move_request_->cancellation_request->issued) {
+        // We have a move request, the samples have been forwarded,
+        // and cancellation is requested but has not yet been issued. Issue a cancel.
+        state.move_request_->cancellation_request->issued = true;
+        if (!arm_conn_->driver->writeTrajectoryControlMessage(
+                urcl::control::TrajectoryControlMessage::TRAJECTORY_CANCEL, 0, RobotReceiveTimeout::off())) {
+            state.move_request_->cancel_error("failed to write trajectory control cancel message to URArm");
+            return event_local_mode_detected_{};
+        }
+    } else if (!state.move_request_->samples.empty() && state.move_request_->cancellation_request) {
+        // We have a move request that we haven't issued but a
+        // cancel is already pending. Don't issue it, just cancel it.
+        std::exchange(state.move_request_, {})->complete_cancelled();
+    } else {
+        // TODO: is it assured that we have positions/velocities here?
+        state.move_request_->write_joint_data(state.ephemeral_->joint_positions, state.ephemeral_->joint_velocities);
+    }
+
+    return std::nullopt;
+}
+
+URArm::state_::state_variant_ URArm::state_::state_controlled_::handle_event(event_connection_lost_) {
+    return state_disconnected_{};
+}
+
+URArm::state_::state_variant_ URArm::state_::state_controlled_::handle_event(event_estop_detected_) {
+    return state_independent_{std::move(arm_conn_), state_independent_::reason::k_estopped};
+}
+
+URArm::state_::state_variant_ URArm::state_::state_controlled_::handle_event(event_local_mode_detected_) {
+    return state_independent_{std::move(arm_conn_), state_independent_::reason::k_local_mode};
+}
+
+template <typename Event>
+URArm::state_::state_variant_ URArm::state_::state_controlled_::handle_event(Event) {
+    return std::move(*this);
+}
+
+URArm::state_::state_independent_::state_independent_(std::unique_ptr<arm_connection_> arm_conn, reason r)
+    : state_connected_(std::move(arm_conn)), reason_(r) {}
+
+std::string_view URArm::state_::state_independent_::name() {
+    using namespace std::literals::string_view_literals;
+    return "independent"sv;
+}
+
+std::string URArm::state_::state_independent_::describe() const {
```

Comment:
Here is where `describe` and such becomes useful: we can give the state a runtime description that includes whether we are estopped, local, or both.